examples/16-3-release-blog-post/forward-ref-example.js

```diff
@@ -1,27 +1,28 @@
 function withTheme(Component) {
-  // Note the second param "ref" provided by React.forwardRef.
-  // We can attach this to Component directly.
-  // highlight-range{1,5}
-  function ThemedComponent(props, ref) {
+  // highlight-next-line
+  function ThemedComponent({forwardedRef, ...rest}) {
     return (
       <ThemeContext.Consumer>
         {theme => (
-          <Component {...props} ref={ref} theme={theme} />
+          // Assign the custom prop "forwardedRef" as a ref
+          // highlight-next-line
+          <Component
+            {...rest}
+            ref={forwardedRef}
+            theme={theme}
+          />
         )}
       </ThemeContext.Consumer>
     );
   }
 
-  // These next lines are not necessary,
-  // But they do give the component a better display name in DevTools,
-  // e.g. "ForwardRef(withTheme(MyComponent))"
-  // highlight-range{1-2}
-  const name = Component.displayName || Component.name;
-  ThemedComponent.displayName = `withTheme(${name})`;
-
-  // Tell React to pass the "ref" to ThemedComponent.
-  // highlight-next-line
-  return React.forwardRef(ThemedComponent);
+  // Note the second param "ref" provided by React.forwardRef.
+  // We can pass it along to ThemedComponent as a regular prop, e.g. "forwardedRef"
+  // And it can then be attached to the Component.
+  // highlight-range{1-3}
+  return React.forwardRef((props, ref) => (
```

Comment:
I don't think we should encourage passing anonymous functions to `React.forwardRef` because it means DevTools won't be able to show a meaningful display name. (This is why the previous example forwarded `displayName`.)

Reply:
DevTools shows `ForwardRef` only if the function is anonymous. If the function has a name, DevTools will show `ForwardRef(ComponentName)` which is much more useful. This is why the original code example I showed had the name copying.

Reply:
> I agree with you. But I think distinguishing between the argument of `React.forwardRef` and Functional Component is more important than that.

I don't see them as mutually exclusive. Why not something like this?

```js
function withTheme(Component) {
  // highlight-next-line
  function ThemedComponent({ forwardedRef, ...rest }) {
    return (
      <ThemeContext.Consumer>
        {theme => (
          // Assign the custom prop "forwardedRef" as a ref
          // highlight-next-line
          <Component {...rest} ref={forwardedRef} theme={theme} />
        )}
      </ThemeContext.Consumer>
    );
  }

  // Note the second param "ref" provided by React.forwardRef.
  // We can pass it along to ThemedComponent as a regular prop, e.g. "forwardedRef"
  // And it can then be attached to the Component.
  // highlight-range{1-3}
  function refForwarder(props, ref) {
    return <ThemedComponent {...props} forwardedRef={ref} />;
  }

  // These next lines are not necessary,
  // But they do give the component a better display name in DevTools,
  // e.g. "ForwardRef(withTheme(MyComponent))"
  // highlight-range{1-2}
  const name = Component.displayName || Component.name;
  refForwarder.displayName = `withTheme(${name})`;

  // highlight-next-line
  return React.forwardRef(refForwarder);
}

// highlight-next-line
const fancyButtonRef = React.createRef();

// fancyButtonRef will now point to FancyButton
// highlight-range{4}
<FancyThemedButton
  label="Click me!"
  onClick={handleClick}
  ref={fancyButtonRef}
/>;
```

> In addition to that, it might cause a problem with static type checking.

Please give an example of this.

Reply:
> I don't see them as mutually exclusive. Why not something like this?

I don't think they are exclusive, but I was not sure what a value `React.forwardRef` accepts?

I thought it accepts a callback function, not Functional Component.
But if the argument is a Functional Component, I think your example makes sense.
Feel free to close this.

> Please give an example of this.

`React.forwardRef` accepts `(props: Props, ref: Ref) => ReactElement`, which seems to be imcompatible with the type definition.

https://github.com/facebook/flow/blob/master/lib/react.js#L132

Actually, the `context` is specified as `any` so the incompatibility does not cause an error though.

Reply:
> I thought it accepts a callback function, not Functional Component.

`forwardRef` _does_ accept a function. To be clear, the only difference in signature between the two is that a "functional component" accepts `(props, context)` as its parameters and the `forwardRef` callback accepts `(props, ref)`.

I can see how this similarity could cause confusion.

In my example above, I used a lowercase name for `refForwarder` to try to make it look a little less like a typical functional component name.

> > In addition to that, it might cause a problem with static type checking.

> Please give an example of this.

I think your original comment about this was based on a misunderstanding. Hopefully my comment above clears things up but if not let me know and I'll elaborate.

Reply:
@bvaughn Thanks! That's good for me. I've updated this PR.